addons/misra.py

```diff
@@ -3185,7 +3187,36 @@ def misra_17_3(self, cfg):
         for w in cfg.clang_warnings:
             if w['message'].endswith('[-Wimplicit-function-declaration]'):
                 self.reportError(cppcheckdata.Location(w), 17, 3)
-
+        for token in cfg.tokenlist:
+            if token.str not in ["while","if"]:
+                continue
+            if token.next.str != "(":
+                continue
+            tok = token.next
+            while (tok != token.next.link):
+                if tok.isName and not tok.function and tok.next.str == "(":
```

Comment:
hmm this is tricky somehow we would like to avoid false positives for standard functions for example:

    if (strlen(str) > 3)

I have the feeling that the "(" of the strlen will have a valuetype.

no warning should be written for `sizeof(int)` the sizeof is a keyword and well the "(" will have a valuetype.

Do not write a warning for this code:

    int (*dostuff)();
    void foo() {
         if (dostuff()) {}  // <- no 17.3 warning
    }